instance/models/load_balancer.py

```diff
@@ -111,6 +145,55 @@ def __init__(self, *args, **kwargs):
     def __str__(self):
         return self.domain
 
+    def modifies_configuration(method):  # nopep8 # pylint: disable=no-self-argument
+        """
+        Flag an internal method as one that modifies the state of load balancer configuration,
+        to allow managing configuration state outside and around the method call.
+        """
+        @functools.wraps(method)
+        def wrapper(self, *args, **kwargs):
+            """
+            Used to wrap the function that modified load balancer configuration state.
+            """
+            try:
+                # Prevent race conditions by updating state only if the LB is actually clean or dirty.
+                updated = LoadBalancingServer.objects.filter(
+                    pk=self.pk,
+                    configuration_state__in=[self.CLEAN, self.DIRTY]
+                ).update(
+                    configuration_state=self.IN_PROGRESS
+                )
+
+                # A race condition or just a normal queuing up of LB reconfiguration requests is handled.
+                if not updated:
+                    raise OtherReconfigurationInProgress
+
+                method(self, *args, **kwargs)  # pylint: disable=not-callable
+            except OtherReconfigurationInProgress:
+                # Prevent a race condition that gets the LB stuck in 'in-progress-and-dirty'
+                # if reconfiguration fails and configuration state is set to dirty, at the same
+                # time that a thread reaches here, by ensuring reconfiguration is really in progress.
+                LoadBalancingServer.objects.filter(
+                    pk=self.pk,
+                    configuration_state=self.IN_PROGRESS
+                ).update(
+                    configuration_state=self.IN_PROGRESS_AND_DIRTY
+                )
+            except ReconfigurationFailed:
+                # If the active reconfiguration failed for any reason, the LB is dirty and ready
+                # for another round of reconfiguration, either by request or periodic task.
+                self.configuration_state = self.DIRTY
```

Comment:
This is a good point. I think we can try and cover more random cases by just trying and using a higher level exception class. That's not the best idea though. If we move it to `finally`, then the next thread that comes in would be able to mark this as being in-progress again.

What we could do to replace the function of the timeout, is just use a timeout context manager, which we can real quickly build, or just use [stopit](https://pypi.python.org/pypi/stopit). It would turn the method call into:

```python
with stopit.ThreadingTimeout(15) as reconfiguration:
    try:
        method(self, *args, **kwargs)
    except ReconfigurationFailed:
        # This particular exception probably doesn't make sense to raise here -- it's what `stopit` expects to change the state, though.
        raise TimeoutException
if reconfiguration.state != reconfiguration.EXECUTED:
    self.configuration_state = self.DIRTY
```

This technique would handle the network timeout case, but not necessarily for exceptions that aren't `ReconfigurationFailed`, so that's one detail to work out.

I know I'm using `self.configuration_state` here. We can just replace that with `update`, and call `save()` at the end, in `finally`, to guarantee the signals. Or whatever approach works best.

Reply:
I believe the only way we can handle the dropped network connection case is by storing a timestamp somewhere – either in Redis or in th DB.  If the network dies, we can't access the DB or Redis anymore, since the network is dead.  So anything we do with `stopit` or context managers or whatever won't be able to save us, since we won't be able to change anything in the DB anyway.

This case actually happens, and the result would be that the LB would get stuck in "in progress" state.  This is a particularly bad failure mode, since it's not really obvious for the user.  You click "Activate" in the UI and your app server gets shown as "active".  You may notice that the feature you meant to deploy does not seem to work yet, but it won't be obvious to you that the UI is lying, and that your app server actually isn't active.  So while this case is easy to fix manually if it happens, it may lead to difficult to debug problems, so I'd prefer if we get this right.

So I really believe we need to use a lock that expires after some time, since this is the only way to avoid that error case.  I think the suggestion I made on Jira using the timestamp and the flag actually works, but it is a bit fiddly.  I will comment on Jira with a suggestion that is easier to understand and implement.